lib/swtpmtest.pm

```diff
@@ -0,0 +1,115 @@
+        ssh_script  => 'ssh_script_1',
+        result_file => '/tmp/swtpm_1',
+    },
+    swtpm_2 => {
+        xml_file    => {uefi => 'swtpm_uefi_2_0.xml', legacy => 'swtpm_legacy_2_0.xml'},
+        version     => '2.0',
+        ssh_script  => 'ssh_script_2',
+        result_file => '/tmp/swtpm_2',
+    },
+};
+
+my $sample_cfg = {
+    uefi   => {vm_name => 'SLES15SP3-UEFI-swtpm',   sample_file => 'swtpm_uefi.xml'},
+    legacy => {vm_name => 'SLES15SP3-LEGACY-swtpm', sample_file => 'swtpm_legacy.xml'},
+};
+
+sub start_swtpm_vm {
+    my ($swtpm_ver, $swtpm_vm_type) = @_;
+    die "invalid swtpm parameter"   unless $guestvm_xml->{$swtpm_ver};
+    die "invalid vm mode parameter" unless $sample_cfg->{$swtpm_vm_type};
+
+    # Copy the sample configuration file and modify it based on the test requirement
+    my $operate_item  = $guestvm_xml->{$swtpm_ver};
+    my $operate_item1 = $sample_cfg->{$swtpm_vm_type};
+    assert_script_run("cd $image_path");
+    assert_script_run("cp $operate_item1->{sample_file} $operate_item->{xml_file}->{$swtpm_vm_type}");
+    assert_script_run(
+"sed -i \"/<\\/devices>/i\\    <tpm model='tpm-tis'>\\n      <backend type='emulator' version='$operate_item->{version}'\\/>\\n    <\\/tpm>\" $operate_item->{xml_file}->{$swtpm_vm_type}"
+    );
+    assert_script_run("virsh define $operate_item->{xml_file}->{$swtpm_vm_type}");
+    assert_script_run("virsh start $operate_item1->{vm_name}");
+}
+
+sub stop_swtpm_vm {
+    my $para = shift;
+    die "invalid vm type parameter" unless $sample_cfg->{$para};
+    my $operate_item = $sample_cfg->{$para};
+    assert_script_run("virsh destroy $operate_item->{vm_name}");
+    if ($para eq "uefi") {
+        assert_script_run("virsh undefine $operate_item->{vm_name} --nvram");
+    }
+    else {
+        assert_script_run("virsh undefine $operate_item->{vm_name}");
+    }
+}
+
+sub swtpm_verify {
+    my $para = shift;
+    die "invalid parameter" unless $guestvm_xml->{$para};
+
+    # Login to the vm and run the commands to check tpm device
+    my $user   = 'root';
+    my $passwd = 'nots3cr3t';
+    my $ip_addr;
+    my $results = script_run("which arp");
+    if ($results) {
+        record_soft_failure("bsc#1180446 - can not install package 'net-tools-deprecated'");
+        $ip_addr = script_output("ip n | grep 192.168 | awk '{print \$1}'");
+    }
+    else {
+        $ip_addr = script_output("arp | grep 192.168 | awk '{print \$1}'");
+    }
+    my $operate_item = $guestvm_xml->{$para};
+    my $result_file  = $operate_item->{result_file};
+    my $ssh_script   = "$image_path/$operate_item->{ssh_script}";
+    assert_script_run("$ssh_script $ip_addr $user $passwd > $result_file");
+    assert_script_run("cat $result_file | grep tpm0");
+    if ($para eq "swtpm_1") {
+        assert_script_run("cat $result_file | grep 'TPM 1.2 Version'");
+    }
+    elsif ($para eq "swtpm_2") {
+        assert_script_run("cat $result_file | grep tpmrm0");
```

Comment:
It will not show up the TPM 2.0, right?

Reply:
let me check, there should be perread cmd

Reply:
pcrread output is filterd